olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -155,18 +153,177 @@ client::CancellableFuture<DataResponse> VersionedLayerClientImpl::GetData(
 
 client::CancellationToken VersionedLayerClientImpl::PrefetchTiles(
     PrefetchTilesRequest request, PrefetchTilesResponseCallback callback) {
-  const int64_t request_key = pending_requests_->GenerateRequestPlaceholder();
+  // Used as empty response to be able to execute initial task
+  using EmptyResponse = Response<PrefetchTileNoError>;
+  using PrefetchResult = PrefetchTilesResult::value_type;
+  using PrefetchResultFuture = std::future<PrefetchResult>;
+  using PrefetchResultPromise = std::promise<PrefetchResult>;
+  using client::CancellationContext;
+  using client::ErrorCode;
+
+  auto catalog = catalog_;
+  auto layer_id = layer_id_;
+  auto settings = *settings_;
+  auto task_scheduler = task_scheduler_;
   auto pending_requests = pending_requests_;
-  auto request_callback = [=](PrefetchTilesResponse response) {
-    if (pending_requests->Remove(request_key) && callback) {
-      callback(std::move(response));
-    }
-  };
+  auto prefetch_repo = prefetch_repo_;
+
+  auto token = AddTask(
+      task_scheduler_, pending_requests_,
+      [=](CancellationContext context) mutable -> EmptyResponse {
+        if (request.GetTileKeys().empty()) {
+          OLP_SDK_LOG_WARNING_F(kLogTag, "PrefetchTiles : invalid request");
```

Comment:
The key will not tell you much if it is empty. It will just print version, layers and levels either way:

```
key=hype-test-prefetch[10/12](1)@4
```

Reply:
Ok, layer is fine too.